File: nim/meetup/meetup.nim
import sequtils, times

type
  Descriptor* = enum
    First, Second, Third, Fourth, Last, Teenth

proc weekDaysInMonth(year: int, month: int, weekDay: WeekDay): seq[DateTime] =
  for day in 1..getDaysInMonth(Month(month), year):
    if getDayOfWeek(day, Month(month), year) == weekDay:
      result.add(initDateTime(day, Month(month), year, 0, 0, 0))

proc formatted(dateTime: DateTime): string =
  dateTime.format("yyyy-MM-dd")

proc meetup*(year: int, month: int, descriptor: Descriptor, weekDay: WeekDay): string =
  let weekDays = weekDaysInMonth(year, month, weekDay)

  case descriptor
  of Teenth: weekDays.filterIt(it.monthday >= 13)[0].formatted
  of First, Second, Third, Fourth: weekDays[descriptor.ord].formatted
  of Last: weekDays[^1].formatted

File: clojure/meetup/src/meetup.clj
(ns meetup
  (:refer-clojure :exclude [zero? range iterate format max min contains?]))

(use 'java-time)

(defn- format-date [date]
  (into [] (as date :year :month-of-year :day-of-month)))

(defn meetup [month year day teenth-day]
  (let [first-day-date (local-date year month)
        first-day-in-month (adjust first-day-date :first-in-month day)
        teenth-date (case teenth-day
                      :first  first-day-in-month
                      :second (plus first-day-in-month (weeks 1))
                      :third  (plus first-day-in-month (weeks 2))
                      :fourth (plus first-day-in-month (weeks 3))
                      :teenth (plus first-day-in-month (weeks (if (>= (as first-day-in-month :day-of-month) 6) 1 2)))
                      :last   (adjust first-day-date :last-in-month day))]
    (format-date teenth-date)))